src/rpc/methods/eth/types.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019-2024 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use super::*;
+
+pub const METHOD_GET_BYTE_CODE: u64 = 3;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct GetBytecodeReturn(pub Option<Cid>);
+
+#[derive(PartialEq, Debug, Deserialize, Serialize, Default, Clone, JsonSchema)]
+pub struct EthAddress(
+    #[schemars(with = "String")]
+    #[serde(with = "crate::lotus_json::hexify_bytes")]
+    pub ethereum_types::Address,
+);
+
+lotus_json_with_self!(EthAddress);
+
+impl EthAddress {
+    pub fn to_filecoin_address(&self) -> anyhow::Result<FilecoinAddress> {
+        if self.is_masked_id() {
+            const PREFIX_LEN: usize = MASKED_ID_PREFIX.len();
+            // This is a masked ID address.
+            let arr = self.0.as_fixed_bytes();
+            let mut bytes = [0; 8];
+            bytes.copy_from_slice(&arr[PREFIX_LEN..]);
+            Ok(FilecoinAddress::new_id(u64::from_be_bytes(bytes)))
+        } else {
+            // Otherwise, translate the address into an address controlled by the
+            // Ethereum Address Manager.
+            Ok(FilecoinAddress::new_delegated(
+                FilecoinAddress::ETHEREUM_ACCOUNT_MANAGER_ACTOR.id()?,
+                self.0.as_bytes(),
+            )?)
+        }
+    }
+
+    // See https://github.com/filecoin-project/lotus/blob/v1.26.2/chain/types/ethtypes/eth_types.go#L347-L375 for reference implementation
+    pub fn from_filecoin_address(addr: &FilecoinAddress) -> anyhow::Result<Self> {
+        match addr.protocol() {
+            Protocol::ID => Ok(Self::from_actor_id(addr.id()?)),
+            Protocol::Delegated => {
+                let payload = addr.payload();
+                let result: Result<DelegatedAddress, _> = payload.try_into();
+                if let Ok(f4_addr) = result {
+                    let namespace = f4_addr.namespace();
+                    if namespace != FilecoinAddress::ETHEREUM_ACCOUNT_MANAGER_ACTOR.id()? {
+                        bail!("invalid address {addr}");
+                    }
+                    let eth_addr: EthAddress = f4_addr.subaddress().try_into()?;
+                    if eth_addr.is_masked_id() {
+                        bail!(
+                            "f410f addresses cannot embed masked-ID payloads: {}",
+                            eth_addr.0
+                        );
+                    }
+                    Ok(eth_addr)
+                } else {
+                    bail!("invalid delegated address namespace in: {addr}")
```

Comment:
nit: you can get rid of this `else` block by returning early in `if let Ok...` and just bailing here.

Reply:
Fixed